File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-0.java
class Solution {
    public int kthSmallest(TreeNode root, int k) {
        java.util.Stack<TreeNode> stack = new java.util.Stack<>();
        TreeNode curr = root;

        while (curr != null || !stack.isEmpty()) {
            // Go to leftmost node
            while (curr != null) {
                stack.push(curr);
                curr = curr.left;
            }

            curr = stack.pop();
            k--;

            if (k == 0) return curr.val;

            curr = curr.right;
        }

        return -1; // should never reach here
    }
}

File: Data Structures & Algorithms/delete-node-in-a-bst/submission-0.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @param {number} key
     * @return {TreeNode}
     */
    deleteNode(root, key) {
        if (!root) return null;

        if (key < root.val) {
            root.left = this.deleteNode(root.left, key);
        } 
        else if (key > root.val) {
            root.right = this.deleteNode(root.right, key);
        } 
        else {
            if (!root.left) return root.right;
            if (!root.right) return root.left;

            let minNode = this.findMin(root.right);

            root.val = minNode.val;

            root.right = this.deleteNode(root.right, minNode.val);
        }

        return root;
    }

    findMin(node) {
        while (node.left) {
            node = node.left;
        }

        return node;
    }
}